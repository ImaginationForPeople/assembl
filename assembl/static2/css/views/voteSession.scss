.votesession-page {
  h1,
  h2,
  h3 {
    margin-bottom: 0;
  }

  .vote-instructions {
    padding: 0 10px;
    font-size: $font-size-m;
    font-family: $font-family4;
    max-width: 780px;
    line-height: 32px;
  }

  .content-section {
    padding: 0;
  }

  .available-tokens {
    margin: 20px 0 40px;

    .category-available-tokens {
      display: flex;
      padding: 0 0 15px;
      max-width: 780px;
      margin-left: -5px;

      h2 {
        margin-top: 7px;
      }

      .tokens {
        padding-left: 100px;
        padding-top: 9px;
      }

      .text {
        margin: 0;
      }

      span {
        font-size: $font-size-xs;
      }
    }
  }

  .available-tokens-sticky {
    position: fixed;
    top: 60px;
    left: 0;
    right: 0;
    z-index: 100;
    border-top: 1px solid $grey1;
    box-shadow: 3px 4px 7px $grey1;

    .separator {
      display: none;
    }

    .category-available-tokens {
      .tokens {
        padding: 20px 0 0 0;
        display: inline-block;
        vertical-align: top;
        text-align: left;

        svg {
          margin: 0 3px;
        }
      }

      .text {
        line-height: 20px;
        margin: 0 0 5px;
        display: inline-block;
        text-align: left;

        .dark-title-2 {
          font-size: $font-size-s;
          line-height: 18px;
        }

        span {
          font-size: $font-size-xs;
        }
      }
    }
  }

  .extra-margin-top {
    margin-top: 400px;
  }

  .tokens > * {
    margin: 7px 3px;
  }

  .theme-box {
    &:first-child {
      margin: -70px 0 0;
    }
  }

  .proposal.row {
    margin: 90px 10px 70px;

    .proposal-title {
      font-family: $font-family;
      margin-top: 0;
      font-size: 22px;
    }

    .text {
      line-height: 24px;
    }

    .tokens-line {
      margin-bottom: 0;

      p {
        margin: 0;
      }

      .tokens {
        margin-left: -5px;
      }
    }

    .tokens > button {
      border: none;
      padding: 0;
      background: none;

      &:focus {
        outline: none;
      }

      &:active {
        box-shadow: none;
      }
    }

    .tokens > .btn-default:hover {
      background-color: transparent;
    }
  }

  .form-actions {
    margin-top: 50px;
    margin-bottom: 100px;
  }

  .link-button {
    border: none;
    font-size: $font-size-xs;
    color: $link-color;
    text-decoration: none;
    background-color: $grey0;
    padding: 15px;

    &:hover {
      color: $link-hover-color;
      text-decoration: none;
    }

    &:focus {
      outline: none;
    }

    &:active {
      box-shadow: none;
    }

    [class^='assembl-icon-']:before,
    [class*=' assembl-icon-']:before {
      font-size: $font-size-xxs;
      margin-left: 5px;
    }
  }

  .link-button:focus,
  .link-button:active,
  .link-button:hover {
    background-color: transparent;
  }

  .expand-row {
    margin: 0 40px;
  }

  .participants-count {
    .assembl-icon-profil:before {
      font-size: 40px;
      margin-bottom: 8px;
    }
  }

  .participants-count__text {
    color: $first-color;
    padding: 0 10px 20px 10px;
  }

  .proposal-results {
    .participants-count {
      border: 1px solid $grey5;
      margin: 20px 0;
      min-height: 120px;
    }

    .doughnut-container {
      text-align: center;
      height: 150px;
      width: 150px;
      padding: 10px 0 0;
      position: relative;
      margin: auto;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 63px;
        left: 53px;
        width: 45px;
        font-size: $font-size-s;

        .vote-totalCount {
          font-size: $font-size-sm;
          font-weight: bold;
          font-family: $font-family3;
        }
      }
    }
  }

  .votes-in-progress {
    margin: 0 40px;
    padding: 0 15px 15px;

    .padding-s {
      padding-left: 5px;
      padding-right: 5px;
    }

    .box {
      border: 1px solid $grey5;
      margin-top: 10px;
      height: 140px;
      text-align: center;
    }

    .tokens-box {
      text-align: center;
    }

    .doughnut-container {
      margin: 0 auto 0 auto;
      height: 80px;
      width: 80px;
      position: relative;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 23px;
        left: 24px;
        width: 34px;
        font-size: $font-size-xxs;

        .vote-totalCount {
          font-size: $font-size-xs;
          font-family: $font-family3;
        }
      }
    }
  }

  .expand-result {
    padding-bottom: 20px;
  }

  .gauge-box {
    padding: 20px 0 0;
  }

  .gauge-box__instructions {
    font-size: $font-size-s;
  }

  .gauge-box__estimate {
    color: $first-color;
  }

  .separator {
    border-bottom: 1px $btn-dark-border-style $grey5;
    max-width: 605px;
    margin-left: -5px;
  }
}

.token-tooltip {
  border-radius: 4px;

  .tooltip-inner {
    background: none;
  }
}

@media screen and (min-width: $screen-sm-min) {
  .votesession-page {
    .proposal.row {
      padding: 15px 5px 15px 0;
      display: flex;
    }

    .participants-count {
      height: inherit;
      padding: 20px 0 0;
    }

    .tokens-box {
      padding: 19px 0 19px 0;
      text-align: center;
    }
  }
}

@media screen and (max-width: $screen-xs-max) {
  .votesession-page {
    .votes-in-progress {
      padding: 0 0 15px 0;

      .box {
        margin-top: 15px;
        height: 160px;
      }
    }

    .available-tokens-sticky {
      display: none;
    }

    .container {
      padding: 0;
      margin: 0;
    }

    .proposal.row {
      padding: 15px 0 15px;
    }

    .participants-count {
      padding: 32px 0 0;
    }

    .tokens-box {
      padding: 10px 0 20px 0;
      text-align: center;
    }

    .section-title {
      display: block;
    }

    .max-container {
      padding-left: 15px;
      padding-right: 15px;
    }
  }
}

@media screen and (min-width: $screen-xl-min) {
  .proposal-results {
    width: 80%;
  }
}

@media screen and (max-width: 660px) {
  .votesession-page {
    .available-tokens {
      .category-available-tokens {
        display: block;

        .tokens {
          padding-left: 10px;
          margin-left: -5px;
        }
      }
    }
  }
}

@media screen and (max-width: 420px) {
  .votesession-page {
    .title-section .dark-title-1 {
      margin: 0 5px;
    }
  }
}
