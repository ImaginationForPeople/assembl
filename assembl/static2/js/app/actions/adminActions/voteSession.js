// @flow
import * as actionTypes from '../actionTypes';
import { createRandomId } from '../../utils/globalFunctions';

export const updateVoteSessionPageTitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_TITLE
});

export const updateVoteSessionPageSeeCurrentVotes = (value: string): actionTypes.UpdateVoteSessionPageSeeCurrentVotes => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SEECURRENTVOTES
});

export const updateVoteSessionPageSubtitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageSubtitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SUBTITLE
});

export const updateVoteSessionPageInstructionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_TITLE
});

export const updateVoteSessionPageInstructionsContent = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsContent => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_CONTENT
});

export const updateVoteSessionPagePropositionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPagePropositionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_PROPOSITIONS_TITLE
});

export const updateVoteSessionHeaderImage = (value: File): actionTypes.UpdateVoteSessionHeaderImage => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_IMAGE
});

export const updateVoteModules = (voteModules: actionTypes.VoteModulesArray): actionTypes.UpdateVoteModules => ({
  voteModules: voteModules,
  type: actionTypes.UPDATE_VOTE_MODULES
});

export const createTokenVoteModule = (id: string): actionTypes.CreateTokenVoteModule => ({
  id: id,
  type: actionTypes.CREATE_TOKEN_VOTE_MODULE
});

export const deleteVoteModule = (id: string): actionTypes.DeleteVoteModule => ({
  id: id,
  type: actionTypes.DELETE_VOTE_MODULE
});

export const createGaugeVoteModule = (id: string): actionTypes.CreateGaugeVoteModule => ({
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_MODULE
});

export const updateTokenVoteExclusiveCategory = (id: string, value: boolean): actionTypes.UpdateTokenVoteExclusiveCategory => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_EXCLUSIVE_CATEGORY
});

export const updateTokenVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateTokenVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_INSTRUCTIONS
});

export const createTokenVoteCategory = (id: string, moduleId: string): actionTypes.CreateTokenVoteCategory => ({
  id: id,
  moduleId: moduleId,
  type: actionTypes.CREATE_TOKEN_VOTE_CATEGORY
});

export const deleteTokenVoteCategory = (moduleId: string, index: number): actionTypes.DeleteTokenVoteCategory => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_TOKEN_VOTE_CATEGORY
});

export const updateTokenVoteCategoryTitle = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryTitle => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_TITLE
});

export const updateTokenTotalNumber = (id: string, value: number, moduleId: string): actionTypes.UpdateTokenTotalNumber => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_TOTAL_NUMBER
});

export const updateTokenVoteCategoryColor = (
  id: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryColor => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_COLOR
});

export const updateGaugeVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateGaugeVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_INSTRUCTIONS
});

export const updateGaugeVoteIsNumber = (id: string, value: boolean): actionTypes.UpdateGaugeVoteIsNumber => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_IS_NUMBER
});

export const updateGaugeVoteNbTicks = (id: string, value: number): actionTypes.UpdateGaugeVoteNbTicks => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_NUMBER_TICKS
});

export const createGaugeVoteChoice = (moduleId: string, id: string): actionTypes.CreateGaugeVoteChoice => ({
  moduleId: moduleId,
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_CHOICE
});

export const deleteGaugeVoteChoice = (moduleId: string, index: number): actionTypes.DeleteGaugeVoteChoice => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_GAUGE_VOTE_CHOICE
});

export const updateGaugeVoteChoiceLabel = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateGaugeVoteChoiceLabel => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_CHOICE_LABEL
});

export const updateGaugeMinimum = (id: string, value: number): actionTypes.UpdateGaugeMinimum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MINIMUM
});

export const updateGaugeMaximum = (id: string, value: number): actionTypes.UpdateGaugeMaximum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MAXIMUM
});

export const updateGaugeUnit = (id: string, value: string): actionTypes.UpdateGaugeUnit => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_UNIT
});

export const updateVoteSessionPage = (value: Object): actionTypes.UpdateVoteSessionPage => ({
  id: value.id,
  titleEntries: value.titleEntries,
  seeCurrentVotes: value.seeCurrentVotes,
  subTitleEntries: value.subTitleEntries,
  instructionsSectionTitleEntries: value.instructionsSectionTitleEntries,
  instructionsSectionContentEntries: value.instructionsSectionContentEntries,
  propositionsSectionTitleEntries: value.propositionsSectionTitleEntries,
  headerImage: value.headerImage,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE
});

export const updateVoteProposals = (voteProposals: actionTypes.VoteProposalsArray): actionTypes.UpdateVoteProposals => ({
  voteProposals: voteProposals,
  type: actionTypes.UPDATE_VOTE_PROPOSALS
});

export const createVoteProposal = (id: string): actionTypes.CreateVoteProposal => ({
  id: id,
  type: actionTypes.CREATE_VOTE_PROPOSAL
});

export const deleteVoteProposal = (id: string): actionTypes.DeleteVoteProposal => ({
  id: id,
  type: actionTypes.DELETE_VOTE_PROPOSAL
});

export const updateVoteProposalTitle = (id: string, locale: string, value: string): actionTypes.UpdateVoteProposalTitle => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_TITLE
});

export const updateVoteProposalDescription = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateVoteProposalDescription => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_DESCRIPTION
});

export const moveProposalUp = (id: string): actionTypes.MoveProposalUp => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_UP
});

export const moveProposalDown = (id: string): actionTypes.MoveProposalDown => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_DOWN
});

export const addModuleToProposal = (
  id: string,
  proposalId: string,
  voteSpecTemplateId: string
): actionTypes.AddModuleToProposal => ({
  id: id,
  proposalId: proposalId,
  voteSpecTemplateId: voteSpecTemplateId,
  type: actionTypes.ADD_MODULE_TO_PROPOSAL
});

export const createVoteProposalAndModules = (id: string) => (dispatch: Function, getState: Function) => {
  dispatch(createVoteProposal(id));
  const { modulesInOrder } = getState().admin.voteSession;
  modulesInOrder.forEach(moduleId => dispatch(addModuleToProposal(createRandomId(), id, moduleId)));
};

export const undeleteModule = (id: string): actionTypes.UndeleteModule => ({
  id: id,
  type: actionTypes.UNDELETE_MODULE
});

/* if a vote spec template params changes, set _hasChanged to true on all its dependencies */
export const markAllDependenciesAsChanged = (id: string): actionTypes.MarkAllDependenciesAsChanged => ({
  id: id,
  type: actionTypes.MARK_ALL_DEPENDENCIES_AS_CHANGED
});

export const setValidationErrors = (id: string, errors: ValidationErrors): actionTypes.SetValidationErrors => ({
  errors: errors,
  id: id,
  type: actionTypes.SET_VALIDATION_ERRORS
});

export const cancelModuleCustomization = (id: string): actionTypes.CancelModuleCustomization => ({
  id: id,
  type: actionTypes.CANCEL_MODULE_CUSTOMIZATION
});