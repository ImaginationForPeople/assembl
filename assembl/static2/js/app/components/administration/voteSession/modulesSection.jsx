// @flow
import * as React from 'react';
import { I18n, Translate } from 'react-redux-i18n';
import { connect } from 'react-redux';
import { Checkbox, SplitButton, MenuItem, Radio } from 'react-bootstrap';
import range from 'lodash/range';

import SectionTitle from '../sectionTitle';
import Helper from '../../common/helper';
import TokensForm from './tokensForm';
import GaugeForm from './gaugeForm';
import {
  createTokenVoteModule,
  createTokenVoteCategory,
  createGaugeVoteModule,
  createGaugeVoteChoice,
  deleteVoteModule,
  undeleteModule,
  updateVoteSessionPageSeeCurrentVotes
} from '../../../actions/adminActions/voteSession';
import { createRandomId } from '../../../utils/globalFunctions';

type ModulesSectionProps = {
  tokenModules: Object,
  gaugeModules: Object,
  editLocale: string,
  toggleModuleCheckbox: Function,
  handleDeleteGauge: (id: string) => void,
  handleGaugeSelectChange: Function,
  handleSeeCurrentVotesChange: Function,
  seeCurrentVotes: boolean
};

const DumbModulesSection = ({
  tokenModules,
  editLocale,
  gaugeModules,
  handleDeleteGauge,
  handleGaugeSelectChange,
  handleSeeCurrentVotesChange,
  seeCurrentVotes,
  toggleModuleCheckbox
}: ModulesSectionProps) => {
  const activeTokenModulesIds = tokenModules.filter(m => !m.get('_toDelete')).map(m => m.get('id'));
  const activeGaugeModulesIds = gaugeModules.filter(m => !m.get('_toDelete')).map(m => m.get('id'));
  const tokenModuleChecked = activeTokenModulesIds.size > 0;
  const gaugeModuleChecked = activeGaugeModulesIds.size > 0;

  return (
    <div className="admin-box">
      <SectionTitle title={I18n.t('administration.voteSession.1')} annotation={I18n.t('administration.annotation')} />
      <div className="admin-content">
        <div className="form-container">
          <div className="vote-modules-form">
            <Checkbox
              id="token-checkbox"
              checked={tokenModuleChecked}
              onChange={() => {
                toggleModuleCheckbox(tokenModuleChecked, tokenModules, 'tokens');
              }}
            >
              <Helper
                label={I18n.t('administration.voteWithTokens')}
                helperUrl="/static2/img/helpers/helper4.png"
                helperText={I18n.t('administration.tokenVoteCheckbox')}
                classname="inline checkbox-title"
              />
            </Checkbox>
            {activeTokenModulesIds.map(id => (
              <TokensForm key={id} id={id} editLocale={editLocale} />
            ))}
            {(tokenModuleChecked || gaugeModuleChecked) && <div className="separator" />}
            <Checkbox
              id="gauge-checkbox"
              checked={gaugeModuleChecked}
              onChange={() => {
                toggleModuleCheckbox(gaugeModuleChecked, gaugeModules, 'gauge');
              }}
            >
              <Helper
                label={I18n.t('administration.voteWithGauges')}
                helperUrl="/static2/img/helpers/helper3.png"
                helperText={I18n.t('administration.gaugeVoteCheckbox')}
                classname="inline checkbox-title"
              />
            </Checkbox>
            {gaugeModuleChecked ? (
              <div className="module-form">
                <div className="flex">
                  <label htmlFor="input-dropdown-addon">
                    <Translate value="administration.gaugeNumber" />
                  </label>
                  <Helper helperUrl="/static2/img/helpers/helper2.jpg" helperText={I18n.t('administration.defineGaugeNumer')} />
                </div>
                <SplitButton
                  className="admin-dropdown"
                  title={activeGaugeModulesIds.size}
                  id="input-dropdown-addon"
                  required
                  onSelect={(eventKey) => {
                    handleGaugeSelectChange(eventKey, activeGaugeModulesIds);
                  }}
                >
                  {range(0, 11).map(value => (
                    <MenuItem key={`gauge-item-${value}`} eventKey={value}>
                      {value}
                    </MenuItem>
                  ))}
                </SplitButton>
              </div>
            ) : null}
            {activeGaugeModulesIds.map((id, index) => (
              <GaugeForm key={id} handleDeleteGauge={handleDeleteGauge} index={index} id={id} editLocale={editLocale} />
            ))}
            {(tokenModuleChecked || gaugeModuleChecked) && (
              <div className="margin-m">
                <label htmlFor="seeCurrentVotes">
                  <Translate value="administration.seeCurrentVotes" />
                </label>
                <Radio
                  id="seeCurrentVotes"
                  onChange={() => {
                    handleSeeCurrentVotesChange(true);
                  }}
                  checked={seeCurrentVotes}
                  name="seeCurrentVotes"
                >
                  <Translate value="administration.resultsVisible" />
                </Radio>
                <Radio
                  onChange={() => {
                    handleSeeCurrentVotesChange(false);
                  }}
                  checked={!seeCurrentVotes}
                  name="seeCurrentVotes"
                >
                  <Translate value="administration.resultsNotVisible" />
                </Radio>
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ({ admin }) => {
  const { modulesInOrder, modulesById, page } = admin.voteSession;
  const { editLocale } = admin;
  return {
    tokenModules: modulesInOrder.map(id => modulesById.get(id)).filter(m => m.get('type') === 'tokens'),
    gaugeModules: modulesInOrder.map(id => modulesById.get(id)).filter(m => m.get('type') === 'gauge'),
    editLocale: editLocale,
    seeCurrentVotes: page.get('seeCurrentVotes')
  };
};

const mapDispatchToProps = (dispatch) => {
  const createGaugeVoteModuleWithChoices = (newId) => {
    dispatch(createGaugeVoteModule(newId));
    dispatch(createGaugeVoteChoice(newId, createRandomId()));
    dispatch(createGaugeVoteChoice(newId, createRandomId()));
  };

  const createTokenVoteModuleWithCategories = (newId) => {
    dispatch(createTokenVoteModule(newId));
    dispatch(createTokenVoteCategory(createRandomId(), newId));
  };

  return {
    toggleModuleCheckbox: (checked, tokenModules, moduleType) => {
      if (!checked) {
        if (tokenModules.size > 0) {
          tokenModules.forEach((m) => {
            dispatch(undeleteModule(m.get('id')));
          });
        } else {
          const newId = createRandomId();
          const dispatchCreateActions = moduleType === 'tokens' ? createTokenVoteModuleWithCategories : createGaugeVoteModuleWithChoices;
          dispatchCreateActions(newId);
        }
      } else {
        tokenModules.forEach((m) => {
          dispatch(deleteVoteModule(m.get('id')));
        });
      }
    },
    handleGaugeSelectChange: (selectedNumber, activeGaugeModulesIds) => {
      if (selectedNumber > activeGaugeModulesIds.size) {
        const numberToCreate = selectedNumber - activeGaugeModulesIds.size;
        for (let i = 0; i < numberToCreate; i += 1) {
          createGaugeVoteModuleWithChoices(createRandomId());
        }
      } else {
        const numberToDelete = activeGaugeModulesIds.size - selectedNumber;
        activeGaugeModulesIds.reverse().forEach((id, index) => {
          if (numberToDelete > index) {
            dispatch(deleteVoteModule(id));
          }
        });
      }
    },
    handleDeleteGauge: (id) => {
      dispatch(deleteVoteModule(id));
    },
    handleSeeCurrentVotesChange: checked => dispatch(updateVoteSessionPageSeeCurrentVotes(checked))
  };
};

export { DumbModulesSection };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DumbModulesSection);