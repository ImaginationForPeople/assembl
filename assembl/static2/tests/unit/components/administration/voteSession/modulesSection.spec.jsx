import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { List, Map } from 'immutable';

import { DumbModulesSection } from '../../../../../js/app/components/administration/voteSession/modulesSection';
import '../../../../helpers/setupTranslations';

describe('ModulesSection component', () => {
  const toggleModuleCheckboxSpy = jest.fn();
  const handleGaugeSelectChangeSpy = jest.fn();
  const handleSeeCurrentVotesChangeSpy = jest.fn();

  it('should render a ModulesSection component without any TokensForm nor GaugeForm', () => {
    const props = {
      tokenModules: List(),
      gaugeModules: List(),
      editLocale: 'fr',
      toggleModuleCheckbox: toggleModuleCheckboxSpy,
      handleGaugeSelectChange: handleGaugeSelectChangeSpy,
      handleSeeCurrentVotesChange: handleSeeCurrentVotesChangeSpy,
      seeCurrentVotes: false
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbModulesSection {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('should render a ModuleSection component with token and gauges', () => {
    const tokenModule = Map({
      _toDelete: true,
      _hasChanged: false,
      _isNew: false,
      id: 'my-token',
      instructionsEntries: List()
    });
    const numGaugeModule = Map({
      _toDelete: false,
      _hasChanged: false,
      _isNew: false,
      id: 'my-num-gauge',
      instructionsEntries: List(),
      max: 11,
      min: 0,
      nbTicks: 4
    });
    const textGaugeModule = Map({
      _toDelete: true,
      _hasChanged: false,
      _isNew: false,
      id: 'my-text-gauge',
      instructionsEntries: List()
    });
    const props = {
      tokenModules: List.of(tokenModule),
      gaugeModules: List.of(numGaugeModule, textGaugeModule),
      editLocale: 'en',
      toggleModuleCheckbox: toggleModuleCheckboxSpy,
      handleGaugeSelectChange: handleGaugeSelectChangeSpy,
      handleSeeCurrentVotesChange: handleSeeCurrentVotesChangeSpy,
      seeCurrentVotes: true
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbModulesSection {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});