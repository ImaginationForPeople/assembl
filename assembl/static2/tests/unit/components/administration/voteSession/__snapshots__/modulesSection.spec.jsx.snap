// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModulesSection component should render a ModuleSection component with token and gauges 1`] = `
<div
  className="admin-box"
>
  <SectionTitle
    annotation="* Fields are required."
    title="Configure the voting modules"
  />
  <div
    className="admin-content"
  >
    <div
      className="form-container"
    >
      <div
        className="vote-modules-form"
      >
        <Checkbox
          bsClass="checkbox"
          checked={false}
          disabled={false}
          inline={false}
          onChange={[Function]}
          title=""
        >
          <Helper
            additionalTextClasses=""
            classname="inline checkbox-title"
            helperText="The token vote module allows you to select propositions proportionnaly. Each participant has a certain amount of tokens et will have to spread them on the different propositions"
            helperUrl="/static2/img/helpers/helper4.png"
            label="Tokens vote"
          />
        </Checkbox>
        Immutable.List [
        ]
        <div
          className="separator"
        />
        <Checkbox
          bsClass="checkbox"
          checked={true}
          disabled={false}
          inline={false}
          onChange={[Function]}
          title=""
        >
          <Helper
            additionalTextClasses=""
            classname="inline checkbox-title"
            helperText="You can choose to have one or several gages"
            helperUrl="/static2/img/helpers/helper3.png"
            label="Gauge(s) vote"
          />
        </Checkbox>
        <div
          className="module-form"
        >
          <div
            className="flex"
          >
            <label
              htmlFor="input-dropdown-addon"
            >
              <Translate
                value="administration.gaugeNumber"
              />
            </label>
            <Helper
              additionalTextClasses=""
              classname=""
              helperText="Define gauges number"
              helperUrl="/static2/img/helpers/helper2.png"
              label=""
            />
          </div>
          <SplitButton
            className="admin-dropdown"
            id="input-dropdown-addon"
            onSelect={[Function]}
            required={true}
            title={1}
          >
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={1}
              header={false}
            >
              1
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={2}
              header={false}
            >
              2
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={3}
              header={false}
            >
              3
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={4}
              header={false}
            >
              4
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={5}
              header={false}
            >
              5
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={6}
              header={false}
            >
              6
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={7}
              header={false}
            >
              7
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={8}
              header={false}
            >
              8
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={9}
              header={false}
            >
              9
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={10}
              header={false}
            >
              10
            </MenuItem>
          </SplitButton>
        </div>
        Immutable.List [
          <Connect(DumbGaugeForm)
            editLocale="en"
            id="my-num-gauge"
            index={0}
          />,
        ]
        <div
          className="margin-m"
        >
          <label
            htmlFor="seeCurrentVotes"
          >
            <Translate
              value="administration.seeCurrentVotes"
            />
          </label>
          <Radio
            bsClass="radio"
            checked={true}
            disabled={false}
            id="seeCurrentVotes"
            inline={false}
            name="seeCurrentVotes"
            onChange={[Function]}
            title=""
          >
            <Translate
              value="administration.resultsVisible"
            />
          </Radio>
          <Radio
            bsClass="radio"
            checked={false}
            disabled={false}
            inline={false}
            name="seeCurrentVotes"
            onChange={[Function]}
            title=""
          >
            <Translate
              value="administration.resultsNotVisible"
            />
          </Radio>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModulesSection component should render a ModulesSection component without any TokensForm nor GaugeForm 1`] = `
<div
  className="admin-box"
>
  <SectionTitle
    annotation="* Fields are required."
    title="Configure the voting modules"
  />
  <div
    className="admin-content"
  >
    <div
      className="form-container"
    >
      <div
        className="vote-modules-form"
      >
        <Checkbox
          bsClass="checkbox"
          checked={false}
          disabled={false}
          inline={false}
          onChange={[Function]}
          title=""
        >
          <Helper
            additionalTextClasses=""
            classname="inline checkbox-title"
            helperText="The token vote module allows you to select propositions proportionnaly. Each participant has a certain amount of tokens et will have to spread them on the different propositions"
            helperUrl="/static2/img/helpers/helper4.png"
            label="Tokens vote"
          />
        </Checkbox>
        Immutable.List [
        ]
        <Checkbox
          bsClass="checkbox"
          checked={false}
          disabled={false}
          inline={false}
          onChange={[Function]}
          title=""
        >
          <Helper
            additionalTextClasses=""
            classname="inline checkbox-title"
            helperText="You can choose to have one or several gages"
            helperUrl="/static2/img/helpers/helper3.png"
            label="Gauge(s) vote"
          />
        </Checkbox>
        Immutable.List [
        ]
      </div>
    </div>
  </div>
</div>
`;
